diff --git a/node_modules/@greenwood/plugin-import-jsx/src/index.js b/node_modules/@greenwood/plugin-import-jsx/src/index.js
index 5fe6e11..736cab1 100644
--- a/node_modules/@greenwood/plugin-import-jsx/src/index.js
+++ b/node_modules/@greenwood/plugin-import-jsx/src/index.js
@@ -4,7 +4,14 @@
  *
  */
 import { generate } from "astring";
-import { parseJsx } from "wc-compiler/src/jsx-loader.js";
+import { parseJsx } from "wc-compiler/jsx-loader.js";
+import { mergeImportMap } from "@greenwood/cli/src/lib/node-modules-utils.js";
+import { parse } from "node-html-parser";
+
+const importMap = {
+  "signal-polyfill": "/node_modules/signal-polyfill/dist/index.js",
+  "wc-compiler/effect": "/node_modules/wc-compiler/src/effect.js"
+};
 
 class ImportJsxResource {
   constructor(compilation, options) {
@@ -12,6 +19,7 @@ class ImportJsxResource {
     this.extensions = ["jsx", "tsx"];
     this.contentType = "text/javascript";
     this.servePage = options.servePages ? "dynamic" : null;
+    this.signals = options.signals;
   }
 
   async shouldServe(url) {
@@ -33,6 +41,42 @@ class ImportJsxResource {
       }),
     });
   }
+
+  async shouldIntercept(url, request, response) {
+    return this.signals && response.headers.get("Content-Type")?.indexOf("text/html") >= 0;
+  }
+
+  async intercept(url, request, response) {
+    const { polyfills } = this.compilation.config;
+    const body = await response.text();
+    let newBody = body;
+
+    if (process.env.__GWD_COMMAND__ === "develop") {
+      newBody = mergeImportMap(newBody, importMap, polyfills.importMaps);
+    }
+
+    const root = parse(newBody);
+    const signalScript = parse(`
+      <script type="module">
+        import { Signal } from 'signal-polyfill';
+        globalThis.Signal = Signal;
+        </script>
+      `,
+    );
+
+    // find the import map script in root and insert the signal polyfill script after it
+    const importMapScript = root.querySelector('script[type="importmap"], script[type="importmap-shim"]');
+
+    if (importMapScript) {
+      importMapScript.after(signalScript);
+    } else {
+      root.querySelector("head").prepend(signalScript);
+    }
+
+    newBody = root.toString();
+
+    return new Response(newBody);
+  }
 }
 
 /** @type {import('./types/index.d.ts').ImportJsxPlugin} */
diff --git a/node_modules/@greenwood/plugin-import-jsx/src/types/index.d.ts b/node_modules/@greenwood/plugin-import-jsx/src/types/index.d.ts
index 1f181ad..0ebc1ba 100644
--- a/node_modules/@greenwood/plugin-import-jsx/src/types/index.d.ts
+++ b/node_modules/@greenwood/plugin-import-jsx/src/types/index.d.ts
@@ -2,6 +2,7 @@ import type { ResourcePlugin } from "@greenwood/cli";
 
 type Options = {
   servePages?: boolean;
+  signals?: boolean;
 };
 
 export type ImportJsxPlugin = (options?: Options) => [ResourcePlugin];
